pkg/component/autoscaling/gardenercustommetrics/component.go

```diff
@@ -0,0 +1,232 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gardenercustommetrics implements the gardener-custom-metrics seed component (aka GCMx).
+// For details, see the GardenerCustomMetrics type.
+package gardenercustommetrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/component/autoscaling/gardenercustommetrics/kubeobjects"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+// ComponentName is the component name.
+const ComponentName = componentBaseName
+
+// GardenerCustomMetrics manages an instance of the gardener-custom-metrics component (aka GCMx). The component is
+// deployed on a seed, scrapes the metrics from all shoot kube-apiserver pods, and provides custom metrics by
+// registering as APIService at the custom metrics extension point of the seed kube-apiserver.
+// For information about individual fields, see the NewGardenerCustomMetrics function.
+type GardenerCustomMetrics struct {
+	namespaceName      string
+	containerImageName string
+	kubernetesVersion  *semver.Version
+
+	seedClient     client.Client
+	secretsManager secretsmanager.Interface
+
+	testIsolation gardenerCustomMetricsTestIsolation // Provides indirections necessary to isolate the unit during tests
+}
+
+// NewGardenerCustomMetrics creates a new GardenerCustomMetrics instance tied to a specific server connection.
+//
+// namespace is where the server-side artefacts (e.g. pods) will be deployed (usually the 'garden' namespace).
+// containerImageName points to the binary for the gardener-custom-metrics pods. The exact version to be used, is
+// determined by contextual configuration, e.g. image vector overrides.
+// If enabled is true, this instance strives to bring the component to an installed, working state. If enabled is
+// false, this instance strives to uninstall the component.
+// seedClient represents the connection to the seed API server.
+// secretsManager is used to interact with secrets on the seed.
+func NewGardenerCustomMetrics(
```

Comment:
Let's follow the [effective go](https://go.dev/doc/effective_go#names) naming guidelines:
```suggestion
func New(
```
Since this is the only "constructor" in this package.

Reply:
+1, the package name is already `gardenercustommetrics`, hence `gardenercustommetrics.New` is better than `gardenercustommetrics.NewGardenerCustomMetrics`.

I already suggested this in https://github.com/gardener/gardener/commit/2317ba0b1e89b0f19d4090c3c72c7365e8d23f17#diff-86c90c029b20e859adf4e91611870a11f58962ee2602a2f439bf99b0a1fe9338R66.

Reply:
Done.